File: BluetoothScanner.cs
// BluetoothScanner.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using InTheHand.Net.Bluetooth;
using InTheHand.Net.Sockets;

namespace BluetoothScanner
{
    class DeviceInfo
    {
        public string Address { get; set; }
        public string Name { get; set; }
        public int RSSI { get; set; }
        public string Timestamp { get; set; }
        public List<string> Services { get; set; }
    }

    class Scanner
    {
        private int timeout;
        private Regex filter;
        private string historyFile;
        private string exportFile;
        private List<DeviceInfo> devices = new List<DeviceInfo>();
        private List<DeviceInfo> history = new List<DeviceInfo>();

        public Scanner(int timeout, string filter, string history, string export)
        {
            this.timeout = timeout;
            if (!string.IsNullOrEmpty(filter))
                this.filter = new Regex(filter, RegexOptions.Compiled);
            this.historyFile = history;
            this.exportFile = export;
            LoadHistory();
        }

        private void LoadHistory()
        {
            if (string.IsNullOrEmpty(historyFile)) return;
            if (File.Exists(historyFile))
            {
                string json = File.ReadAllText(historyFile);
                // Deserialize (Newtonsoft.Json or System.Text.Json) - skip for brevity
            }
        }

        private void SaveHistory()
        {
            if (string.IsNullOrEmpty(historyFile)) return;
            // Serialize and write
        }

        private bool MatchesFilter(string name)
        {
            return filter == null || (name != null && filter.IsMatch(name));
        }

        public async Task ScanAsync()
        {
            Console.WriteLine($"Scanning for {timeout} seconds...");
            var client = new BluetoothClient();
            var devicesFound = new List<BluetoothDeviceInfo>();

            var task = Task.Run(() =>
            {
                client.DiscoverDevicesAsync(255, false, true, true);
                // Actually we need to use the old DiscoverDevices method (blocking) for simplicity
                devicesFound = client.DiscoverDevices(255, false, true, true).ToList();
            });

            if (await Task.WhenAny(task, Task.Delay(timeout * 1000)) == task)
            {
                // completed early
            }
            else
            {
                // timeout
                client.CancelDiscover();
            }

            foreach (var d in devicesFound)
            {
                if (MatchesFilter(d.DeviceName))
                {
                    var info = new DeviceInfo
                    {
                        Address = d.DeviceAddress.ToString(),
                        Name = d.DeviceName ?? "Unknown",
                        RSSI = d.Rssi, // may not be available
                        Timestamp = DateTime.UtcNow.ToString("o"),
                        Services = new List<string>()
                    };
                    devices.Add(info);
                }
            }

            Console.WriteLine($"Found {devices.Count} devices:");
            foreach (var d in devices)
                Console.WriteLine($"  {d.Name} ({d.Address}) RSSI: {d.RSSI} dBm");

            if (!string.IsNullOrEmpty(historyFile))
            {
                history.AddRange(devices);
                SaveHistory();
                Console.WriteLine("History updated.");
            }
        }

        public static async Task Main(string[] args)
        {
            int timeout = 10;
            string filter = null, history = null, export = null;
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--timeout": timeout = int.Parse(args[++i]); break;
                    case "--filter": filter = args[++i]; break;
                    case "--history": history = args[++i]; break;
                    case "--export": export = args[++i]; break;
                }
            }
            var scanner = new Scanner(timeout, filter, history, export);
            await scanner.ScanAsync();
            // Export CSV if needed
        }
    }
}

File: bluetooth_scanner.cpp
// bluetooth_scanner.cpp
#include <iostream>
#include <vector>
#include <string>
#include <regex>
#include <thread>
#include <chrono>
#include <fstream>
#include <json/json.h> // using jsoncpp
#include <bluetooth/bluetooth.h>
#include <bluetooth/hci.h>
#include <bluetooth/hci_lib.h>
#include <unistd.h>

struct DeviceInfo {
    std::string address;
    std::string name;
    int rssi;
    std::string timestamp;
    std::vector<std::string> services;
};

class Scanner {
private:
    int timeout;
    std::regex filter;
    bool hasFilter;
    std::string historyFile;
    std::string exportFile;
    std::vector<DeviceInfo> devices;
    std::vector<DeviceInfo> history;

public:
    Scanner(int t, const std::string& f, const std::string& h, const std::string& e)
        : timeout(t), hasFilter(!f.empty()), historyFile(h), exportFile(e) {
        if (hasFilter) filter = std::regex(f, std::regex::icase);
        loadHistory();
    }

    void loadHistory() {
        if (historyFile.empty()) return;
        std::ifstream ifs(historyFile);
        if (ifs) {
            Json::Value root;
            ifs >> root;
            // parse history
        }
    }

    void saveHistory() {
        if (historyFile.empty()) return;
        Json::Value root(Json::arrayValue);
        for (auto& d : history) {
            Json::Value item;
            item["address"] = d.address;
            item["name"] = d.name;
            item["rssi"] = d.rssi;
            item["timestamp"] = d.timestamp;
            Json::Value svcs(Json::arrayValue);
            for (auto& s : d.services) svcs.append(s);
            item["services"] = svcs;
            root.append(item);
        }
        std::ofstream ofs(historyFile);
        ofs << root;
    }

    bool matchesFilter(const std::string& name) {
        return !hasFilter || std::regex_search(name, filter);
    }

    void scan() {
        std::cout << "Scanning for " << timeout << " seconds..." << std::endl;
        int dev_id = hci_get_route(nullptr);
        int sock = hci_open_dev(dev_id);
        if (sock < 0) {
            perror("hci_open_dev");
            return;
        }
        int len = 8;
        int max_rsp = 255;
        char buf[HCI_MAX_EVENT_SIZE];
        struct hci_inquiry_req ir = { 0 };
        ir.dev_id = dev_id;
        ir.num_rsp = max_rsp;
        ir.length = timeout;
        ir.flags = IREQ_CACHE_FLUSH;

        auto rsps = (inquiry_info*)malloc(max_rsp * sizeof(inquiry_info));
        int count = hci_inquiry(dev_id, len, max_rsp, nullptr, &rsps, IREQ_CACHE_FLUSH);
        if (count < 0) perror("hci_inquiry");
        else {
            for (int i = 0; i < count; i++) {
                char addr[19];
                ba2str(&rsps[i].bdaddr, addr);
                char name[248] = {0};
                if (hci_read_remote_name(sock, &rsps[i].bdaddr, sizeof(name), name, 0) < 0)
                    strcpy(name, "Unknown");
                if (matchesFilter(name)) {
                    DeviceInfo info;
                    info.address = addr;
                    info.name = name;
                    info.rssi = rsps[i].rssi;
                    info.timestamp = std::to_string(time(nullptr));
                    devices.push_back(info);
                }
            }
        }
        close(sock);
        free(rsps);

        std::cout << "Found " << devices.size() << " devices:" << std::endl;
        for (auto& d : devices) {
            std::cout << "  " << d.name << " (" << d.address << ") RSSI: " << d.rssi << " dBm" << std::endl;
        }
        if (!historyFile.empty()) {
            history.insert(history.end(), devices.begin(), devices.end());
            saveHistory();
            std::cout << "History updated." << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    int timeout = 10;
    std::string filter, history, exportFile;
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--timeout" && i+1 < argc) timeout = std::stoi(argv[++i]);
        else if (arg == "--filter" && i+1 < argc) filter = argv[++i];
        else if (arg == "--history" && i+1 < argc) history = argv[++i];
        else if (arg == "--export" && i+1 < argc) exportFile = argv[++i];
    }
    Scanner scanner(timeout, filter, history, exportFile);
    scanner.scan();
    return 0;
}
